feat: Add LLM Security Notebooks - Prompt, ReAct, RAG (#977)
GoogleCloudPlatform/generative-ai: gemini/security/README.md

Reviewer: Don't refer to the notebooks specifically as "Colabs" just call them Notebooks since they can be opened in other methods.

Change the name of the files to be all lowercase with no special characters except for `_`

Move all ipynb files directly into the security directory.

The files in `data/` should be added to Google Cloud Storage instead of in the repository. I will upload them and share the link here. Have the users download the files from GCS or access them remotely through the Gemini API.

Remove all internal-only links, this repository is public.
- You can add the slide decks to GCS or change them to markdown.

Author: Many thanks for the review, I will add them to PR

# Attacks and Mitigation Labs 
 **Prompt, RAG and ReAct**

>  This comprehensive learning CoLabs dives deep into the critical world of LLM security, equipping you with the knowledge and skills to build, deploy, and maintain secure and trustworthy AI solutions. From understanding the latest attack vectors to implementing cutting-edge defenses, this CoLabs is your guide to navigating the evolving landscape of LLM security.

Reviewer: ```suggestion
 These comprehensive learning notebooks dive deep into the critical world of LLM security, equipping you with the knowledge and skills to build, deploy, and maintain secure and trustworthy AI solutions. From understanding the latest attack vectors to implementing cutting-edge defenses, these notebooks are your guide to navigating the evolving landscape of LLM security.
```

- LLM Prompt attacks and mitigation 

  CoLab [GenAI_&_LLM_Security_(v2_Gemini)_For_Developers_Prompt_Attacks,_Mitigation_Techniques_&_Protection_Strategies.ipynb](colabs/GenAI_&_LLM_Security_(v2_Gemini)_For_Developers_Prompt_Attacks,_Mitigation_Techniques_&_Protection_Strategies.ipynb)
  - Simple prompt design prompt design
  - Antipatterns on prompt design with PII data and secrets
  - Prompt Attacks:
    - Data Leaking
    - Data Leaking with Transformations
    - Modifying the Output (Jailbreaking)
    - Hallucinations
    - Payload Splitting
    - Virtualization
    - Obfuscation
    - Multimodal Attacks (Image, PDF & Video)
    - Model poisioning
  - Protections & Mitigations with:
    - Data Loss Prevention
    - Natural Language API (Category Check, Sentiment Analysis)
    - Malware checking
    - LLM validation (Hypothesis Validation, DARE, Strict Input Validation with Random Token)
    - Responsible AI Safety filters
    - Embeddings

> - ReAct and RAG attacks and mitigation options [GenAI_and_LLM_Security_ReAct_and_RAG_attacks_&_mitigations.ipynb](GenAI_and_LLM_Security_ReAct_and_RAG_attacks_&_mitigations.ipynb)

Reviewer: ```suggestion
- [ReAct and RAG attacks and mitigation options](GenAI_and_LLM_Security_ReAct_and_RAG_attacks_&_mitigations.ipynb)
```

  - Simple ReAct applicaiton design 
  - Attacks using ReAct Agent 
  - Mitigation using ReAct Agent information
  - Simple RAG applicaiton 
  - Attacks and mitigation for RAG applications

> 

Reviewer: Please run a markdown formatter on this.
